coreax/solvers/coresubset.py

```diff
@@ -435,29 +435,53 @@ def reduce(
         data. The pivot sampling and diagonal updating steps are repeated until
         :math:`M` points have been selected.
 
+        .. note::
+            The RPCholesky algorithm sometimes converges before reaching
+            ``self.coreset_size``. In this case, the other coreset points are chosen
+            uniformly at random, avoiding repetition.
+
+
         :param dataset: The dataset to reduce to a coresubset
         :param solver_state: Solution state information, primarily used to cache
             expensive intermediate solution step values.
         :return: a refined coresubset and relevant intermediate solver state information
         """
         x = dataset.data
+        num_data_points = len(x)
         if solver_state is None:
             gramian_diagonal = jax.vmap(self.kernel.compute_elementwise)(x, x)
         else:
             gramian_diagonal = solver_state.gramian_diagonal
-        initial_coresubset = _initial_coresubset(0, self.coreset_size, dataset)
+        # Initialise with an index outside of the possible coreset indices
+        initial_coresubset = _initial_coresubset(
+            num_data_points + 1, self.coreset_size, dataset
+        )
         coreset_indices = initial_coresubset.unweighted_indices
-        num_data_points = len(x)
 
         def _greedy_body(
             i: int, val: tuple[Array, Array, Array]
         ) -> tuple[Array, Array, Array]:
             """RPCholesky iteration - Algorithm 1 of :cite:`chen2023randomly`."""
             residual_diagonal, approximation_matrix, coreset_indices = val
             key = jr.fold_in(self.random_key, i)
+
+            # Fall back on uniform probability if residual_diagonal is invalid
```

Comment:
The problem with adding noise to simulate random choice is that the residual diagonal becomes `nan` anyway, so this stops working. Instead, we check if the diagonal is all 0 or has nans and then force random choice between the unpicked points.